panic/log/src/lib.rs

```diff
@@ -4,15 +4,41 @@
 #![deny(missing_docs, missing_debug_implementations)]
 #![no_std]
 
-extern crate alloc;
-use alloc::string::ToString;
+use core::fmt::Write;
 use core::panic::PanicInfo;
+use mc_sgx_io::WriteBuffer;
+use mc_sgx_sync::Mutex;
+
+/// A buffer for building up the panic message.
+/// We avoid allocating when handling the panic as failure to allocate could be
+/// the cause of the panic.
+static MESSAGE_BUFFER: Mutex<WriteBuffer> = Mutex::new(WriteBuffer::new());
 
 #[panic_handler]
 fn panic(info: &PanicInfo) -> ! {
-    // Ignore the result, we're already panicking we can't really do much if
-    // `stderr_write_all()` fails
-    let _ = mc_sgx_io::stderr_write_all(info.to_string().as_bytes());
-
+    log_panic_info(info);
     loop {}
```

Comment:
The more I'm thinking about this, the more I think there should be a specific `panic()` OCALL instead of `stderr_write_all()`. As it's written right now, this will write the panic message, then go into an infinite loop when there's a panic in the enclave.

Unfortunately, there's nothing on the untrusted side to trigger a userspace panic either, so the app will just hang, and calls into the enclave will just error. The behavior we really want is for a panic in the enclave to cause the app to die, so the `panic()` function should do write into the buffer, then do an OCALL to a panic-handler-proxy in untrusted.

Reply:
Are you thinking panic handler in untrusted or "abort()" in trusted?

Reply:
it's hard to fully parse, https://download.01.org/intel-sgx/sgx-linux/2.18/docs/Intel_SGX_Developer_Reference_Linux_2.18_Open_Source.pdf#page=136
It looks like abort kills the one thread, but also puts the enclave in a state where only the current threads can continue and then the whole thing needs to be restarted.

Letting the untrusted side handle it might allow an application to isolate just that one thread, but I'm not sure how much urts lets one into mucking with the threads

Reply:
Added task to #17